students/mt-desta/lab1/README.md

```diff
@@ -0,0 +1,107 @@
+
+<img src="img/image1.png">
+
+####  <p style='text-aligh: justify'> 1.3 The first dataset is made up of **clusters with jumpers**. As clearly seen on the plot we have one or two points that connect the areas with dense consentration of points. As for the second dataset, it has no clear shape but I assume is made up of **overlapping clusters**.  </p>
+
+#### <p style='text-aligh: justify'> 1.4 Though it is not clear, for the first dataset my assumption is at least 5 clusters by looking at the graph. As for the second dataset, there seem to be 2 or 3 areas with consentration but I cannot be so sure.</p>
+
+#### 2. I have implented a hierarchical algorithm based on Agglomerative clustering. [my_hierarchy.py](./source/my_hierarchy.py) contains the code for this implementation. It takes the number of desired clusters, linkage method and a dataframe as arguements. It outputs the linkage matrix and labels. I have implemented a function tha takes these and plots a dendrogram.
+
+<img src="img/image2.png">
+<img src="img/image3.png">
+
+#### 2.1 Dendrograms for the two datasets
+Dendrogram plotted using the implemented hierarchical algorithm.
+
+<img src="img/image4.png">
+
+Dendrogram using the reference algorithm.
+
+<img src="img/image5.png">
+
+As we can see in the figures, we have similar results.
+        
+#### 2.2 Optimum number of Clusters
+
+#### <p style='text-aligh: justify'> The optimum number of clusters is calculated using maximum increment R_t for hierarchical clustering based on Ward distance. </p>
+
+The results on the first dataset
+```
+Optimum number of clusters for dataset 1 : 4
+Max increment distance: 1668.6530013777597
+```
+The results on the second dataset
+```
+Optimum number of clusters for dataset 2 : 4
+Max increment distance: 29.204787217021345
+```
+    
+####    According to the results for the Mall Customers Dataset 4 clusters seems to be the appropriate number. For the Iris dataset, I chose also 4 clusters as per the results. 
+
+#### 3. I have implemented [EM algorithm](./source/em.py) and [DBSCAN algorithm](./source/dbscan.py).
+
+<img src="img/image6.png" width='400'>
+<img src="img/image7.png" width='400'>
+
+<img src="img/image8.png">
+<img src="img/image9.png">
+
+#### 4. The following metrics were applied on those algorithms.
+####    * Based on the clustering results, calculated the average intra-cluster distance;
+####    * Based on the clustering results, calculated the average inter cluster distance;
+####    * measured the clustering speed;
+
+### For Mall Customers: 
+
+| | Metric |	Implemented Algorithms |	Library Algorithm |
+| - |----- | ----------------------- | ------------------ |
+|1 |	Average Inter-Cluster Distance (Agglomerative) |	59.040257 |	58.271701 |
+| 2 |	Average Inter-Cluster Distance (EM) |	52.191699 |	56.166705 |
+| 3 |	Average Inter-Cluster Distance (DBSCAN) |	53.652448 |	46.076062 |
+| 4 |	Average Intra-Cluster Distance (Agglomerative) |	304.787110 |	24.506988 |
```

Comment:
Why are the values ​​so different?

Reply:
I thought I wrote better code but it turns out there was a bug in the code. The library code returns cluster labels not distances, so my calculations were wrong. Resolved the issues.